### ISSUE TITLE
remove menu separator from sitemap xml

### ISSUE THREAD

**Issue by thomaszbz** [mblaschke/TYPO3-metaseo#24]
Menu separator "pages" occur in the sitemap xml although there is no content for the entry.

How to reproduce:
- website with existing menu, one entry per page
- add new page of type "menu separator" (which is rendered as LI without a A href=... in the menu)

Result
- There is a new entry for the menu separator page in the sitemap xml (which is wrong).
- The corresponding request for the url returns 404 status code (which is correct!).

Suggested fix:
- Remove menu separator items from sitemap xml or allow configuration (globally or per page)

Tested on pre-2.0.0 ( 598e3f66e14d24f6d04fa60f9f179835cde5a590 )

**Comment by thomaszbz** [comment]
executing garbage collection task and sitemap.xml task, plus clearing the typo3 caches, does not solve this: The menu separator items reappear in the sitemap.xml. Google webmaster tools is reporting a bunch of links with status code 404.

**Comment by thomaszbz** [comment]
This is a duplicate of https://forge.typo3.org/issues/46003

**Comment by thomaszbz** [comment]
There is a global array $PAGES_TYPE which contains the page types (from my 6.2 installation):

    Array
    (
        [3] => Array ()
        [4] => Array ()
        [6] => Array ( // backend section
                [type] => web
                [allowedTables] => *
            )
        [7] => Array ()
        [199] => Array // menu separator
            (
                [type] => sys
            )
        [254] => Array // folder
            (
                [type] => sys
                [allowedTables] => *
            )
        [255] => Array // recycler
            (
                [type] => sys
                [allowedTables] => *
            )
        [default] => Array
            (
                [type] => web
                [allowedTables] => pages,sys_category,sys_file_reference,sys_file_collection,pages_language_overlay,tt_content,sys_template,sys_domain,backend_layout,tx_rtehtmlarea_acronym,sys_note
                [onlyAllowedTables] => 0
            )
    )

The corresponding indexes are stored in pages.doktype in the database. They can be retrieved via 

    $GLOBALS['TSFE']->page->doktype

However, when checking against blacklisted entries, metaseo uses

    //metaseo/Classes/Hook/SitemapIndexHook.php, line 520
    // Check for type blacklisting
    if (in_array($GLOBALS['TSFE']->type, $this->typeBlacklist) ) {
        return FALSE;
    }

which is wrong for checking the page type constants.

Plus, the DOKTYPE Constants from

    \TYPO3\CMS\Frontend\Page\PageRepository

could be used for BC instead of hardcoding them.

References:
http://docs.typo3.org/typo3cms/CoreApiReference/PageTypes/Index.html

**Comment by thomaszbz** [comment]
debug output of $GLOBALS['TSFE'] from inside the SitemapIndexHook class:

    TYPO3\CMS\Frontend\Controller\TypoScriptFrontendController Object
    (
        [id] => 73
        [type] => 0
        [cHash] =>
        [no_cache] =>
        [rootLine] => Array
            (
                [1] => Array
                    (
                        < parent page
                    )
    
                [0] => Array
                    (
                        < last one is the root page
                    )
    
            )
    
        [page] => Array  //current page
            (
                [uid] => 73
                [pid] => 2
                [title] => Services
                [doktype] => 199

**Comment by thomaszbz** [comment]
There is another problem in metaseo/Classes/Hook/SitemapIndexHook.php:

It seems as if the sitemap table is filled whenever a request is done:
- The requested current page itsself is added via hook_indexContent() which executes addPageToSitemapIndex() and then checks the requested page for non-blacklisted doktype via checkIfCurrentPageIsIndexable()
- hook_linkParse() is executed whenever a link is parsed while rendering the requested page. The rendered link, however is not the requestet page. Anyways, the doktype is checked again via checkIfCurrentPageIsIndexable(), which is checking the doktype of the requested page and not the doktype of the page of the parsed link.

    $GLOBALS['TSFE']->page->['doktype']

cannot be used here, because $GLOBALS['TSFE'] is the context of the request, not the context of the parsed link processed via hook_linkParse().

**Comment by thomaszbz** [comment]
$pObj, the parameter of hook_linkParse(&$pObj) contains the following content for a menu separator:

    array(4) {
      ["conf"]=>
      &array(2) {
        ["parameter"]=>
        string(2) "82"
        ["linkAccessRestrictedPages"]=>
        bool(false)
      }
      ["linktxt"]=>
      &string(1) "|"
      ["finalTag"]=>
      &string(47) "<a href="x/abc/">"
      ["finalTagParts"]=>
      &array(4) {
        ["aTagParams"]=>
        string(0) ""
        ["url"]=>
        string(36) "x/abc/"
        ["targetParams"]=>
        string(0) ""
        ["TYPE"]=>
        string(4) "page"
      }
    }

The TYPE "page" is right, because actually, a menu separator is a page. What is needed would be the doktype of that page. That information is missing in the $pObj. I could ony imagine to get the doktype via database query by retrieving it via the page uid which resides in "parameters". In the example, the uid would be 82.

**Comment by thomaszbz** [comment]
Performance: checkIfCurrentPageIsIndexable() has no parameters and is executed for every parsed link in the same way via hook_linkParse(&$pObj).

That said, hook_linkParse(&$pObj) should not execute the current checkIfCurrentPageIsIndexable() at all.

A new function is required:

- Create a function checkIfPageIsIndexable($uid = null) to be used via hook_linkParse(&$pObj)
- This function could lookup the doktype and return false for the blacklisted doktypes
- To be investigated: which parts of checkIfCurrentPageIsIndexable() apart of blacklisted doktypes need to be triggered by checkIfPageIsIndexable()?

**Comment by thomaszbz** [comment]
$this->pageIndexFlag is in the context of the current request. If this is true for the current request, then checkIfCurrentPageIsIndexable() returns true for all the links parsed.

As a result, checkIfCurrentPageIsIndexable() is not checking anything for all the pages that come via hook_linkParse(&$pObj).

I would say, $this->pageIndexFlag could be removed completely, because the "caching" is only valid ONE time for the current request. After that, the behaviour is even wrong for other pages than the currently requested one. There is no caching effect at all as long as the hook_indexContent() is executed once only.

**Comment by thomaszbz** [comment]
Some refactoring should be done. It's pretty much the same what hook_indexContent() and hook_linkParse(&$pObj) should be doing, however, the implementation must be different most of the time. And where they are currently using the same code, it's mostly wrong.

Instead, two classes should be used for the two hooks. Both could inherit central functionality from a new abstract class which could look like this

    abstract class SitemapIndexHook ... {
       protected $processedCache; // this object could be a singleton as there's two objects later for two classes
        //<- class variables/settings go here
        protected function check1();
        protected function check2();
        protected function extractPageData();
        protected function addPageToSitemapIndex($pageData) {
            //stub
        }
        protected function indexUrl() {  //to be executed by the hook
            //process $this->check1();
            //process $this->check2();
            $this->addPageToSitemapIndex($this->extractPageData());
        }
    }

Then, there would be a clear structure for both of the hooks, still allowing different implementations.

**Comment by thomaszbz** [comment]
Workaround to get rid of menu separators in the sitemap.xml:

- In the backend, click metaseo in the left side
- click the menu separator page in the page tree
- click "searchengine information" in the top dropdown
- in the table, set "excluded" to "true".

**Comment by mblaschke** [comment]
Found a good solution in eafb6be2c9e48e91139099a42c94c93dc9a613bb
Can you try it again?

Thanks for your contribution and your good feedback :)

**Comment by thomaszbz** [comment]
For testing 786801717c9853bd0a13b57962a65d86359b6f15

- I removed the workaround (set separator pages visible to search engines again
- flushed caches
- ran the garbage collection
- clicked around in the frontend

Result: no separator pages occured in the sitemap.xml

For a user's point of view, the bug seems to be fixed now. Congrats for that!

From a developer's perspective there is still some refactoring left. Just reread my analysis.

**Comment by mblaschke** [comment]
Really thanks for all your infomations.

There seems to be some very ugly confusion and i also got confused:

$GLOBALS['TSFE']->page['doktype'] is ne Page doktype (Page, Folder, Mount, Shortcut, Separator...)

$GLOBALS['TSFE']->type is ne typeNum from the current PAGE object from your SetupTS configuration

I've splitted configuration now which hopefully fixes the confusion - anyway the blacklist was really wrong and in this point i get really confused:

- the PageRepository constants must be checked against the doktype
- the sitemap.* and robots.txt types must be checked against type (typenum)

That was a really really ugly bug and should be fixed now

I also splitted hook from page and link parsing but because of performance issue i don't watch to fetch the doktype from parsed link. It's not the best solution but i don't want to slow down a webpage and risk some bad seo rating because of slow loading times.
The filter in sitemap output and in garbage collection should be no big issue.

Next up: refactor sitemap indexing hooks :)
